adafruit_ble_berrymed_pulse_oximeter/adafruit_ble_transparent_uart.py

```diff
@@ -86,14 +92,14 @@ def readline(self):
         return self._rx.readline()
 
     @property
-    def in_waiting(self):
+    def in_waiting(self) -> int:
         """The number of bytes in the input buffer, available to be read."""
         return self._rx.in_waiting
 
-    def reset_input_buffer(self):
+    def reset_input_buffer(self) -> None:
         """Discard any unread characters in the input buffer."""
         self._rx.reset_input_buffer()
 
-    def write(self, buf):
+    def write(self, buf: bytes) -> None:
```

Comment:
This could actually be more generalized, I am pretty sure that this can be generalized to anything that uses the buffer protocol (https://docs.circuitpython.org/en/latest/docs/reference/glossary.html?highlight=buffer%20protocol#term-buffer-protocol).  Specifically, it doesn't even need to be mutable, so we can type this as CircuitPython's `ReadableBuffer` type:

```suggestion
def write(self, buf: ReadableBuffer) -> None:
```

Don't forget to import it after the `typing` import:

```python
from circuitpython_typing import ReadableBuffer
```

You'll need to make sure that this library (`adafruit-circuitpython-typing`) gets added to the `requirements.txt` so it's downloaded during the CI, as well as added to `setup.py` so it gets installed whenever its downloaded via `pip` whenever its used on Blinka.  The library is actually now included with Blinka version 7.2.3, so you can add a minimum version to both `requirements.txt` and `setup.py` to guarantee it's downloaded too:

For `requirements.txt`, you can modify the Blinka like to include the minimum version:

```
Adafruit-Blinka>=7.2.3
```

You can do that same trick in `setup.py` in the `install_requires` variable (https://github.com/adafruit/Adafruit_CircuitPython_BLE_BerryMed_Pulse_Oximeter/blob/96d17e7a03bef0c3eba0bba8d7ad3f313194aa4a/setup.py#L36):

```
    install_requires=["Adafruit-Blinka>=7.2.3", "adafruit-circuitpython-ble"],
```

Reply:
I think I got it all, submitting changes shortly.

This might be the most helpful comment I've ever gotten in a PR, thank you tekktrik!

Reply:
My pleasure!  Glad it was helpful and not word spaghetti! :laughing: